modules/media_duplicate_validation/src/Plugin/MediaDuplicateValidation/ColorMean.php

```diff
@@ -0,0 +1,382 @@
+<?php
+
+namespace Drupal\media_duplicate_validation\Plugin\MediaDuplicateValidation;
+
+use Drupal\media\Entity\Media;
+use Drupal\media\MediaInterface;
+use Drupal\media_duplicate_validation\Plugin\MediaDuplicateValidationBase;
+
+/**
+ * Compares an image to other images.
+ *
+ * To do this we resize all images to a 25x25 grayscale image. Then comparing
+ * the average color value of each row and column we collect a subset of similar
+ * images. Then we compare pixel by pixel and with the thresholds we can mark
+ * each one as similar or not.
+ *
+ * @MediaDuplicateValidation(
+ *   id = "color_mean"
+ * )
+ */
+class ColorMean extends MediaDuplicateValidationBase {
+
+  /**
+   * Database table used for this plugin.
+   */
+  const DATABASE_TABLE = 'media_validate_color_mean';
+
+  /**
+   * Percent different each color can be to be considered the same.
+   */
+  const PIXEL_TOLLERANCE = 25;
+
+  /**
+   * Total percent that is different but still considered similar.
+   */
+  const IMAGE_TOLLERANCE = 10;
+
+  /**
+   * The dimensions to resize the images to compare against.
+   */
+  const RESIZE_DIMENSION = 25;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSimilarItems(MediaInterface $entity) {
+    if (!($file = $this->getFile($entity, ['image']))) {
+      return [];
+    }
+
+    // The file failed because its either not an image or its a gif.
+    if (!($image_colors = $this->getColorData($file->getFileUri()))) {
+      return [];
+    }
+
+    $similar_media = [];
+
+    /** @var \Drupal\media\Entity\Media $entity */
+    foreach ($this->getCloseMedia($entity, $image_colors) as $similar_entity) {
+
+      $similar_file = $this->getFile($similar_entity, ['image']);
+      $file_likeness = $this->getLikeness($file->getFileUri(), $similar_file->getFileUri());
+
+      // The percent likeness is within the threshold we have defined.
+      if (100 - $file_likeness <= self::IMAGE_TOLLERANCE) {
+
+        // In case multiple images have the same likeness, change the percent
+        // just a tiny bit so it can still be part of the set.
+        while (isset($similar_media["$file_likeness"])) {
```

Comment:
This kind of feels like a bit of a cheat. I get you do this for the sorting below but it does bring up the problem of how to sort equal things. Is there another parameter we want to use to evaluate whether or not something is more like the uploaded image. Perhaps file name? But then what if both file names are the same? This might be a box of spiders to worry about later.